modules/aws/lambda.go

```diff
@@ -2,51 +2,136 @@ package aws
 
 import (
 	"encoding/json"
+	"errors"
 	"fmt"
 
 	"github.com/aws/aws-sdk-go/service/lambda"
 	"github.com/gruntwork-io/terratest/modules/testing"
 	"github.com/stretchr/testify/require"
 )
 
+// LambdaOptions contains additional parameters for InvokeFunctionWithParams().
+// It contains a subset of the fields found in the lambda.InvokeInput struct.
+type LambdaOptions struct {
+	// FunctionName is a required field containing the lambda function name.
+	FunctionName *string
+
+	// InvocationType can be one of "RequestResponse" or "DryRun".
+	//    * RequestResponse (default) - Invoke the function synchronously.
+	//    Keep the connection open until the function returns a response
+	//    or times out.
+	//
+	//    * DryRun - Validate parameter values and verify that the user or
+	//    role has permission to invoke the function.
+	InvocationType *string
+
+	// Lambda function input; will be converted to JSON.
+	Payload interface{}
+}
+
+// LambdaOutput contains the output from InvokeFunctionWithParams().  The
+// fields may or may not have a value depending on the invocation type and
+// whether an error occurred or not.
+type LambdaOutput struct {
+	// If present, indicates that an error occurred during function execution.
+	// Error details are included in the response payload.
+	FunctionError *string
```

Comment:
Why include this in the output rather than returning it as an `error` as the function did before? It seems more Go idiomatic to wrap this in an `error`, rather than embedding it in the output struct.

Reply:
[But that's how the AWS SDK represents a FunctionError in their InvokeOutput struct.](https://github.com/aws/aws-sdk-go/blob/v1.38.4/service/lambda/api.go#L11905-L11927).

`InvokeFunctionE()`'s `FunctionError` struct is equivalent to `InvokeFunctionWithParamsE()`'s `InvokeOutput`.  Which will be confusing, but my contention is that `InvokeFunctionE()`'s `FunctionError` is a misnomer.  For example, for DryRun, a non-nil `FunctionError` is not an error as it contains output, e.g., a status code.

The fact that any other kind of `FunctionError` from `InvokeFunctionE()` is treated as an error is why I needed to make these code changes in the first place.

Reply:
Since the AWS API is very... odd here... Wouldn't it make more sense to have the new `InvokeFunctionWithParamsE` check for `FunctionError` and handle it accordingly? E.g., If a Dry Run `FunctionError` isn't _really_ an error, then handle it accordingly?

Reply:
I'm very confused and I'm not sure what you're asking or if I do understand it, if I agree with you.  

`InvokeFunctionWithParamsE()` returns `LambdaOutput` and `error`:

- `error` represents the error that could occur in the process of invoking the lambda, e.g., bad arguments passed to `InvokeFunctionWithParmsE()` itself. 

- `LambdaOutput` doesn't "judge" the output.  It's not trying to guess what is or is not a valid response from the lambda.  Where I work, we invoke lambdas that have do not return a payload and issue AWS SDK calls to verify the expected side effects take place, so we're only interested in the `error` return and maybe the status code in `LambdaOutput`.  

I don't want to call `LambdaOutput` `FunctionError` because that's assuming the contents represent an error, which is not always the case.  The output can have no payload and a status code of "200" and that's success, not an error.  For DryRun, there's no payload and a status code of "204".  That's success.

If the Lambda function returns a status code the caller is not expecting, then `LambdaOutput.FunctionError` provides details that can be logged for future debugging.

If what you're saying is that you want `LambdaOutput.FunctionError` to be of type `Error`, I can do that.

Reply:
> I'm very confused and I'm not sure what you're asking or if I do understand it, if I agree with you.
> 
> `InvokeFunctionWithParamsE()` returns `LambdaOutput` and `error`:
> 
> * `error` represents the error that could occur in the process of invoking the lambda, e.g., bad arguments passed to `InvokeFunctionWithParmsE()` itself.
> * `LambdaOutput` doesn't "judge" the output.  It's not trying to guess what is or is not a valid response from the lambda.  Where I work, we invoke lambdas that have do not return a payload and issue AWS SDK calls to verify the expected side effects take place, so we're only interested in the `error` return and maybe the status code in `LambdaOutput`.
> 
> I don't want to call `LambdaOutput` `FunctionError` because that's assuming the contents represent an error, which is not always the case. The output can have no payload and a status code of "200" and that's success, not an error. For DryRun, there's no payload and a status code of "204". That's success.
> 
> If the Lambda function returns a status code the caller is not expecting, then `LambdaOutput.FunctionError` provides details that can be logged for future debugging.
> 
> If what you're saying is that you want `LambdaOutput.FunctionError` to be of type `Error`, I can do that.

I think the mental model I have is that the signature is roughly:

```go
InvokeFunctionWithParams(...) (*LambdaOutput, error)
```

Where `LambdaOutput` is:

```go
type LambdaOutput struct {
  Payload []byte
  StatusCode *int64
}
```

Any error `InvokeFunctionWithParams` hits, whether while invoking the Lambda, or returned in `out.FunctionError`, is returned in the second `error` param. The rest goes into `LambdaOutput`. The idea of having to check for errors in the `error` param _and_ in `LambdaOutput` strikes me as very confusing.

Reply:
Thank you for the write up; that cleared up the confusion.  I think the confusion was due to `FunctionError` being the name of a struct and a field in another struct; I think we weren't talking about the same `FunctionError`.

I removed `FunctionError` from `LambdaOutput`.  I had to restore the original code for `InvokeFunctionE()` because the values returned from `InvokeFunctionWithParams()` no longer distinguish between parameter errors and invocation errors, and `InvokeFunctionE()` needs to know the difference to determine whether `FunctionError{}` should be populated.